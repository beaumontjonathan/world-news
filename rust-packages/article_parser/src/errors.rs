use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
