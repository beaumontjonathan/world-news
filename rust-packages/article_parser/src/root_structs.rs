use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
