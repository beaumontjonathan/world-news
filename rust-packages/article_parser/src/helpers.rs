use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
