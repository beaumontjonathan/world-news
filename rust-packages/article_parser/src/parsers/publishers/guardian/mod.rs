use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
use crate::helpers::{
  AttrValueOrNone, ChildrenHelpers, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};

use crate::my_selectors;
use chrono::{DateTime, Utc};
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_VIDEO_PRECHECK => "body > div > article[id=\"article\"][itemtype=\"http://schema.org/VideoObject\"].content--media--video",
  TWITTER_TITLE_SELECTOR => "head > meta[name=\"twitter:text:title\"]",
  DATE_UPDATED_SELECTOR => "meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "meta[property=\"article:published_time\"]",
  HEADER => "div.content__main-column > header",
  HEADLINE => "h1.content__headline",
  DESCRIPTION => "div.content__standfirst",
  FIGURE_HEADER => "figure.media-primary",
  IMG => "img",
  FIGCAPTION => "figcaption.caption",
  ARTICLE_CONTENT => "div.content__article-body",
  ELEMENT_IMAGE => "figure.element-image",
  ELEMENT_INTERACTIVE => "figure.element-interactive",
  VIDEO_FIGURE => "figure.element[data-atom-type=\"media\"]",
  YOUTUBE_IFRAME => "div.youtube-media-atom__iframe",
}

type Result<T> = std::result::Result<T, InternalParseError>;

pub fn parse(document: Html) -> ParseResult {
  // PRECHECKS
  match document.select(&ARTICLE_VIDEO_PRECHECK).next() {
    None => Ok(()),
    Some(_) => Err(InternalParseError::unknown(
      "Article type: video article, not yet supported",
    )),
    // Some(_) => Err(InternalParseError::precondition(
    //   "Article type: video article, not yet supported",
    // )),
  }?;

  let header = document.select_unique(&HEADER)?;
  let article_content = document.select_unique(&ARTICLE_CONTENT)?;

  // HEADLINE
  let headline = {
    let h1s = header
      .select(&HEADLINE)
      .map(|h1| h1.full_fixed_html_text())
      .collect::<Vec<_>>();
    match h1s.len() {
      0 => Err(InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector: &HEADLINE,
      })),
      1 => Ok(h1s.into_iter().next().unwrap()),
      _ => {
        if h1s.windows(2).all(|w| w[0] == w[1]) {
          Ok(h1s.into_iter().next().unwrap())
        } else {
          Err(InternalParseError::unknown(
            "Unknown headline type - too many <h1>s with differing text",
          ))
        }
      }
    }
  }?;

  // TWITTER HEADLINE
  let twitter_headline = document
    .select_unique_attr(&TWITTER_TITLE_SELECTOR, "content")
    .ok();

  // DESCRIPTION
  let description = header.select_unique(&DESCRIPTION)?.full_fixed_html_text();

  // THUMBNAIL
  // todo - implement
  let thumbnail = {
    let figure = header.select_unique(&FIGURE_HEADER)?;
    let img = figure.select_unique(&IMG)?;
    let alt = img.attr_value_or_none("alt");
    let url = img.attr_value_or_none("src");
    let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
    Image { alt, url, caption }
  };

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_content
    .children_all()
    .into_iter()
    .filter(|el| ELEMENT_IMAGE.matches(&el) || ELEMENT_INTERACTIVE.matches(&el))
    .map(|figure| {
      if ELEMENT_IMAGE.matches(&figure) {
        let img = figure.select_unique(&IMG)?;
        let alt = img.attr_value_or_none("alt");
        let url = img.attr_value_or_none("src");
        let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
        Ok(Image { alt, url, caption })
      } else if ELEMENT_INTERACTIVE.matches(&figure) {
        // unimplemented!();
        Ok(Image {
          alt: None,
          url: None,
          caption: None,
        })
      } else {
        panic!("how are you here?");
      }
    })
    .collect::<Result<Vec<_>>>()?;

  // VIDEOS
  let videos = article_content
    .children_all()
    .into_iter()
    .filter(|figure| VIDEO_FIGURE.matches(&figure))
    .map(|figure| {
      let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
      let youtube_id = figure.select_unique_attr(&YOUTUBE_IFRAME, "data-asset-id")?;
      let url = Some(format!("https://www.youtube.com/watch?v={}", youtube_id));

      Ok(Video {
        alt: None,
        url,
        caption,
      })
    })
    .collect::<Result<Vec<_>>>()?;

  // BODY
  let body = article_content
    .children_all()
    .into_iter()
    .filter_map(|el| {
      if el.value().name() == "p" || el.value().name() == "ul" {
        Some(el.full_fixed_html_text())
      } else {
        None
      }
    })
    .collect::<Vec<String>>()
    .into_iter()
    .join_with_newline();

  // DATE_UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.with_timezone(&Utc);

  //DATE_PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.with_timezone(&Utc);

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  // todo - error nicely?
  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_05() {
    let document = read_html_doc("test-05.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_08() {
    let document = read_html_doc("test-08.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
