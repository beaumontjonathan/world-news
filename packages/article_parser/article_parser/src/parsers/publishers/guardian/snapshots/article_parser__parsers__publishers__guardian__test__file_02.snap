---
source: src/parsers/publishers/guardian/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "Bernie Sanders reassesses campaign after Biden builds formidable lead",
            twitter_headline: None,
            description: "The Vermont senator’s presidential campaign has seen little but bad news recently as it admits ‘losing the battle over electability’",
            thumbnail: Image {
                alt: Some(
                    "Senator Bernie Sanders arrives before votes in the Senate, in Washington on Wednesday. Sanders denied reports that he would suspend his presidential campaign.",
                ),
                url: Some(
                    "https://i.guim.co.uk/img/media/5e9aaa288e39ff81e53163cb17a34b449422cda2/0_127_3842_2307/master/3842.jpg?width=300&quality=85&auto=format&fit=max&s=9891b95954463768161fdd8d511f87c8",
                ),
                caption: Some(
                    "Senator Bernie Sanders arrives before votes in the Senate, in Washington on Wednesday. Sanders denied reports that he would suspend his presidential campaign. Photograph: Michael Reynolds/EPA",
                ),
            },
            categories: [],
            images: [
                Image {
                    alt: Some(
                        "Joe Biden and Bernie Sanders greet each other with their elbows at the last Democratic presidential debate, in Washington. Sanders has been forced to cancel the rallies that are a pillar of his campaign.",
                    ),
                    url: Some(
                        "https://i.guim.co.uk/img/media/d40ee9b857eda90c3c895169619554b440585287/0_200_6000_3600/master/6000.jpg?width=300&quality=85&auto=format&fit=max&s=858a1c4d1413336817a74cb7816ed438",
                    ),
                    caption: Some(
                        "Joe Biden and Bernie Sanders greet each other with their elbows at the last Democratic presidential debate, in Washington. Sanders has been forced to cancel the rallies that are a pillar of his campaign. Photograph: Gabriella Demczuk For Cnn/EPA",
                    ),
                },
            ],
            videos: [],
            body: "The Vermont senator Bernie Sanders got thumped in the latest round of Democratic presidential primary contests and, once again, Sanders and his team are reassessing their path forward.\nIt’s becoming a familiar cycle: former vice-president Joe Biden beats Sanders, the darling of progressive grassroots voters; the Sanders campaign concedes the situation isn’t ideal, and spends some time reassessing. That’s what happened earlier this month after the Super Tuesday contests on 3 March, and this Tuesday night and Wednesday morning.\nOn Wednesday Sanders, the campaign said, would first vote on coronavirus legislation in the Senate and then later in the day he would fly with his wife Jane back to Vermont where he would have conversations with supporters.\n“The next primary contest is at least three weeks away. Senator Sanders is going to be having conversations with supporters to assess his campaign,” Sanders’ campaign manager, Faiz Shakir, said in a statement. “In the immediate term, however, he is focused on the government response to the coronavirus outbreak and ensuring that we take care of working people and the most vulnerable.”\nThe Sanders campaign also sent out an email to supporters with a similarly gloomy tone.\n“No sugarcoating it, last night did not go the way we wanted,” Shakir wrote in that email. “And while our campaign has won the battle of ideas, we are losing the battle over electability to Joe Biden .”\nThe email did not include a usual link to make a donation to the campaign. Instead it linked to Sanders’ response on how the federal government should be responding to the coronavirus pandemic, a signal that the campaign is at least slightly inching away from a relentless campaign mode.\nThere were three contests on Tuesday night and Biden beat Sanders in all of them. More ominous for Sanders, Biden beat him by double digits in all three states. Biden won decisively by about 14 points in Illinois.\nIn 2016, Sanders narrowly lost to former secretary of state Hillary Clinton there by less than three percentage points. In Florida, Biden beat Sanders by almost 40 percentage points. In Arizona, Sanders lost to Biden by about the same margin he lost to Clinton in 2016.\nThe prospect of Sanders dropping out is so strong that for a few moments on Wednesday a report about Sanders suspending campaign ads on Facebook was misread as Sanders suspending his campaign.\nThe next set of contests are in states that were more favorable to Sanders in 2016 –Hawaii and Alaska, in particular. But the larger problem for the Vermont senator is the delegate math. As of Wednesday Biden led Sanders by 1,153 pledged delegates to 874. Even before Tuesday, Sanders would have to have won the lion’s share of delegates in coming contests.\nThe Biden campaign has been hesitant to declare victory outright but has argued publicly and behind the scenes that Biden’s delegate advantage is unassailable.\nIn a conference call with donors ahead of the Tuesday’s primary results, Greg Schultz, a top adviser for the campaign, said the Biden team felt they had a “very clear delegate trajectory that would require Sanders to so overwhelmingly out-perform” in the next few primary contests, according to audio obtained by the Guardian.\nOn Wednesday morning, Unite the Country, a Super Pac supporting Biden, released a memo going even further.\n“Last night’s overwhelming victories in Arizona, Illinois and Florida are likely to add at least 150 delegates to Joe Biden’s already commanding delegate lead. Even without Ohio voting , Joe Biden effectively ended the race last night,” the memo said.\n“He will be the Democratic nominee, and for Unite the Country, the focus now turns to November . The turnout generated by the vice-president continues to surprise everyone. Despite the virus , total turnout in Arizona and Florida exceeded 2016 numbers, and just like in previous states this primary season, there was particularly robust turnout in the types of suburban communities where we need to show strength this fall.”\nMore concerning for Sanders and his progressive movement within the primary contest are the limitations the coronavirus outbreak is putting on the campaign. Because of the pandemic, Sanders’ campaign cannot hold the energetic rallies that are a pillar of the campaign, Democratic strategist Julian Mulvey, who advised Sanders in 2016, said.\n“Part of the difficult challenge now is they can’t hold rallies. The rallies have always been, in one sense, the central part of the campaign,” Mulvey said. “It’s direct connection with Bernie and the grassroots and people. It’s him getting on the ground, delivering his 90-minute stump speech – delivering his giant stump speech. I think it’s very important for the energy on the ground, it’s very important for Bernie.”\nBut Sanders allies and strategists from the progressive wing of the party still see value in the Vermont senator continuing the campaign.\n“In this moment of national crisis, Bernie’s platform is more salient than ever: Republicans and centrist Democrats alike are suddenly echoing his calls for free healthcare, expanded paid sick leave, and unemployment insurance,” progressive Democratic strategist Monica Klein said in an email to the Guardian on Wednesday.\nShe added: “We’ve seen from exit polls that Bernie’s economic and healthcare platforms are widely supported by Democratic voters – so whoever we end up with as our nominee, the Democratic ticket needs to embrace the bold economic solutions that Bernie has championed since day one.”",
            date_updated: 2020-03-18T21:11:28Z,
            date_published: 2020-03-18T18:25:55Z,
        },
    ),
)
